src/include/btree.i

```diff
@@ -981,8 +981,7 @@ __wt_page_can_evict(WT_SESSION_IMPL *session, WT_PAGE *page, int check_splits)
 	 * been written in the checkpoint, leaving the checkpoint inconsistent.
 	 */
 	if (btree->checkpointing &&
-	    (__wt_page_is_modified(page) ||
-	    F_ISSET(mod, WT_PM_REC_MULTIBLOCK))) {
+	    (__wt_page_is_modified(page) || F_ISSET(mod, WT_PM_REC_MASK))) {
```

Comment:
Ouch. How did we not see that before now?

We could acquire a RW lock on the parent page (reconciliation holds it for a long time, eviction would hold it for a very short period)?

Alternatively, in the non-split case, it's an atomic update, something like this:
```
diff --git a/src/evict/evict_page.c b/src/evict/evict_page.c
index 892d5b4..5d8860a 100644
--- a/src/evict/evict_page.c
+++ b/src/evict/evict_page.c
@@ -162,7 +162,7 @@ __wt_evict_page_clean_update(WT_SESSION_IMPL *session, WT_REF *ref)
 static int
 __evict_page_dirty_update(WT_SESSION_IMPL *session, WT_REF *ref, int exclusive)
 {
-       WT_ADDR *addr;
+       WT_ADDR *addr, *prevaddr;
        WT_PAGE *parent;
        WT_PAGE_MODIFY *mod;
 
@@ -215,11 +215,6 @@ __evict_page_dirty_update(WT_SESSION_IMPL *session, WT_REF *ref, int exclusive)
                        WT_RET(__wt_split_multi(session, ref, exclusive));
                break;
        case WT_PM_REC_REPLACE:                         /* 1-for-1 page swap */
-               if (ref->addr != NULL && __wt_off_page(parent, ref->addr)) {
-                       __wt_free(session, ((WT_ADDR *)ref->addr)->addr);
-                       __wt_free(session, ref->addr);
-               }
-
                /*
                 * Update the parent to reference the replacement page.
                 *
@@ -232,8 +227,14 @@ __evict_page_dirty_update(WT_SESSION_IMPL *session, WT_REF *ref, int exclusive)
                mod->mod_replace.size = 0;
 
                __wt_ref_out(session, ref);
+               prevaddr = ref->addr;
                ref->addr = addr;
                WT_PUBLISH(ref->state, WT_REF_DISK);
+
+               if (prevaddr != NULL && __wt_off_page(parent, prevaddr)) {
+                       __wt_free(session, prevaddr->addr);
+                       __wt_free(session, prevaddr);
+               }
                break;
        WT_ILLEGAL_VALUE(session);
        }
```